app/uk/gov/hmrc/transitmovementsrouter/controllers/UpscanResponseParser.scala

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.transitmovementsrouter.controllers
+
+import play.api.Logging
+import play.api.libs.json.JsValue
+import play.api.mvc.BaseController
+import uk.gov.hmrc.transitmovementsrouter.models.responses.UpscanResponse
+
+trait UpscanResponseParser {
+  self: BaseController with Logging =>
+
+  def parseAndLogUpscanResponse(responseBody: JsValue): Option[UpscanResponse] =
```

Comment:
I think better to move this method inside controller and change method name to validateUpscanResponse.

Reply:
I'm non-committal about it. I can see why we'd have it as a trait, if only for testing it'll potentially make things easier. I'd probably put it in the controller myself, but it's not a dealbeaker for me. 

Do we need the `BaseController` self type?

Also, rather than returning an option, we know that this will return an appropriate error message at some point. I think this should be returning `EitherT[Future, PresentationError, UpscanResponse]` and return a bad request if we  got a bad response from Upscan.

Leave actually processing the response for success/failure for now. I envisage that this function will be split into two, validate (create the business object) and process (use the business object), the latter probably going into a service.

Reply:
@drnaylor-oc  The request will coming from upscan, I thought that our response back to them would be irrelevant, which is why I didn't implement an error in an eitherT (I did at first, then realised there's no point?) but perhaps I'm misunderstanding something

Reply:
>  The request will coming from upscan, I thought that our response back to them would be irrelevant, which is why I didn't implement an error in an eitherT (I did at first, then realised there's no point?) but perhaps I'm misunderstanding something

If they've sent us something that doesn't fit their schema/contract (broken Json, or a field that no longer exists, for example), we should tell them by sending a 400. If we get a "file failed" callback with a valid failure schema then _that's_ still a 2xx, because we'll handle the failure successfully.

> I made the BaseController a self type because of the issue you pointed out regarding this functionality belonging in the controller/presentation layer.

Okay! I just thought it might be a bit of a pain from a testing view, but that makes sense. Not too bothered about removing it.